collect_app/src/test/java/org/odk/collect/android/formentry/QuitFormDialogFragmentTest.java

```diff
@@ -0,0 +1,85 @@
+package org.odk.collect.android.formentry;
+
+import android.content.DialogInterface;
+import android.widget.TextView;
+
+import androidx.appcompat.app.AlertDialog;
+import androidx.fragment.app.FragmentActivity;
+import androidx.fragment.app.FragmentManager;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.odk.collect.android.R;
+import org.odk.collect.android.formentry.saving.FormSaveViewModel;
+import org.odk.collect.android.support.RobolectricHelpers;
+import org.robolectric.RobolectricTestRunner;
+import org.robolectric.shadows.ShadowDialog;
+
+import static android.view.View.GONE;
+import static android.view.View.VISIBLE;
+import static junit.framework.TestCase.assertTrue;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+import static org.odk.collect.android.support.RobolectricHelpers.mockViewModelProvider;
+import static org.robolectric.Shadows.shadowOf;
+
+@RunWith(RobolectricTestRunner.class)
+public class QuitFormDialogFragmentTest {
+
+    private FragmentActivity activity;
+    private FragmentManager fragmentManager;
+    private QuitFormDialogFragment dialogFragment;
+    private FormSaveViewModel formSaveViewModel;
+
+    @Before
+    public void setup() {
+        activity = RobolectricHelpers.createThemedActivity(FragmentActivity.class);
+        fragmentManager = activity.getSupportFragmentManager();
+        dialogFragment = new QuitFormDialogFragment();
+
+        formSaveViewModel = mockViewModelProvider(activity, FormSaveViewModel.class).get(FormSaveViewModel.class);
+    }
+
+    @Test
+    public void shouldShowCorrectButtons() {
+        dialogFragment.show(fragmentManager, "TAG");
+        AlertDialog dialog = (AlertDialog) ShadowDialog.getLatestDialog();
+
+        assertNotNull(dialog);
+        assertThat(dialog.getButton(DialogInterface.BUTTON_POSITIVE).getVisibility(), equalTo(GONE));
+        assertThat(dialog.getButton(DialogInterface.BUTTON_NEGATIVE).getVisibility(), equalTo(VISIBLE));
+        assertThat(dialog.getButton(DialogInterface.BUTTON_NEGATIVE).getText(),
+                equalTo(activity.getString(R.string.do_not_exit)));
+    }
+
+    @Test
+    public void shouldShowCorrectTitle() {
+        dialogFragment.show(fragmentManager, "TAG");
+        AlertDialog dialog = (AlertDialog) ShadowDialog.getLatestDialog();
+        String title =  formSaveViewModel.getFormName() == null ? activity.getString(R.string.no_form_loaded)
```

Comment:
The problem with this is that it's unclear what the test wants the title to be. For "unit" tests like this you want you test to be as in charge of the dependencies as possible. Given the ViewModel is a mock you can set the value that will be returned from `getFormName` like so:

```
when(formSaveViewModel.getFormName()).thenReturn("blah");
```

This means you could write a test for both the null and non null case. Here I I'd imagine you're only testin the null case as the mock will return null from `getFormName()`.

Reply:
Oh alright yes. We should test both the cases